src/libraries/Microsoft.PowerFx.Core/IR/Visitors/DependencyVisitor.cs

```diff
@@ -0,0 +1,357 @@
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.PowerFx.Core.IR.Nodes;
+using Microsoft.PowerFx.Core.IR.Symbols;
+using Microsoft.PowerFx.Types;
+using static Microsoft.PowerFx.Syntax.PrettyPrintVisitor;
+
+namespace Microsoft.PowerFx.Core.IR
+{
+    // IR has already:
+    // - resolved everything to logical names.
+    // - resolved implicit ThisRecord
+    internal class DependencyVisitor : IRNodeVisitor<DependencyVisitor.RetVal, DependencyVisitor.DependencyContext>
+    {
+        // Track reults.
+        public DependencyInfo Info { get; private set; } = new DependencyInfo();
+
+        public DependencyVisitor()
+        {
+        }
+
+        public override RetVal Visit(TextLiteralNode node, DependencyContext context)
+        {
+            return null;
+        }
+
+        public override RetVal Visit(NumberLiteralNode node, DependencyContext context)
+        {
+            return null;
+        }
+
+        public override RetVal Visit(BooleanLiteralNode node, DependencyContext context)
+        {
+            return null;
+        }
+
+        public override RetVal Visit(DecimalLiteralNode node, DependencyContext context)
+        {
+            return null;
+        }
+
+        public override RetVal Visit(ColorLiteralNode node, DependencyContext context)
+        {
+            return null;
+        }
+
+        public override RetVal Visit(RecordNode node, DependencyContext context)
+        {
+            // Read all the fields. The context will determine if the record is referencing a data source
+            foreach (var kv in node.Fields)
+            {
+                AddField(context, context.ScopeType?.TableSymbolName, kv.Key.Value);
+
+                // Reset the context for field values in case we are operating in a write context.
+                kv.Value.Accept(this, new DependencyContext());
+            }
+
+            return null;
+        }
+
+        public override RetVal Visit(ErrorNode node, DependencyContext context)
+        {
+            return null;
+        }
+
+        public override RetVal Visit(LazyEvalNode node, DependencyContext context)
+        {
+            return node.Child.Accept(this, context);
+        }
+
+        private readonly Dictionary<int, FormulaType> _scopeTypes = new Dictionary<int, FormulaType>();
+
+        public override RetVal Visit(CallNode node, DependencyContext context)
+        {
+            if (node.Scope != null)
+            {
+                // Functions with more complex scoping will be handled by the function itself.
+                var arg0 = node.Args[0];
+                _scopeTypes[node.Scope.Id] = arg0.IRContext.ResultType;
+            }
+
+            node.Function.ComposeDependencyInfo(node, this, context);
+
+            return null;
+        }
+
+        public override RetVal Visit(BinaryOpNode node, DependencyContext context)
+        {
+            node.Left.Accept(this, context);
```

Comment:
>context [](http://example.com/codeflow?start=35&length=7)

We don't need to pass the WriteState flag through here. 

In all cases, a write should be immediately obvious at the mutation function be looking at its arg's imediate types and not require a visitor.